sdk/azidentity/errors_test.go

```diff
@@ -0,0 +1,59 @@
+//go:build go1.18
+// +build go1.18
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azidentity
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"testing"
+)
+
+// testing our custom Error interface for the authentication failed error type
+func TestAuthenticationFailedErrorInterface(t *testing.T) {
+	urlString := "http://www.test.com"
+	resBodyString := "Test Error Response Body"
+	url, err := url.Parse(urlString)
+	if err != nil {
+		t.Fatal("URL parse failed")
+	}
+	req := &http.Request{Method: "POST", URL: url}
+	res := &http.Response{
+		Status:     "400 Bad Request",
+		StatusCode: 400,
+		Body:       ioutil.NopCloser(bytes.NewBufferString(resBodyString)),
+		Request:    req,
+	}
+	err = newAuthenticationFailedError(credNameAzureCLI, "error message", res)
+	var e *AuthenticationFailedError
+	if !errors.As(err, &e) {
+		t.Fatal("expected AuthenticationFailedError")
+	}
+	if e.RawResponse == nil {
+		t.Fatal("expected a non-nil RawResponse")
+	}
+	errMsg := e.Error()
```

Comment:
If the constructor returned an `AuthenticationFailedError` wrapped by another type of error, that would be a bug. However, `errors.As` would return true because it would find an `AuthenticationFailedError` by unwrapping `err`. You can use a type assertion to check whether the thing returned by the constructor is itself an `AuthenticationFailedError`:
```suggestion
	if e, ok := err.(*AuthenticationFailedError); ok {
		if e.RawResponse == nil {
			t.Fatal("RawResponse shouldn't be nil")
		}
	} else {
		t.Fatalf("expected AuthenticationFailedError, got %T", err)
	}
	errMsg := err.Error()
```

Reply:
switched to type assertion in commit [8c8df0d](https://github.com/Azure/azure-sdk-for-go/pull/18779/commits/8c8df0dfcfa7f0d4a61965f749ae8cfa92a9fa4c)